Fix initializer naming at torch.onnx.ExportOutput.save_model_with_external_data

thiagocrepaldi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105002

This PR is only relevant for the Fake tensor Mode ONNX export. For the conventional export, everything is unchanged.

* An optional `rename_initializer=False` argument is added to an internal function `torch/onnx/_internal/fx/serialization.py::save_model_with_external_data` which is used by the public API `ExportOutput.save`. 
* The default behavior (`rename_initializer=False`) is meant to be used by public API `torch.onnx.dynamo_export` with the default Dynamo-based FX tracer (`DynamoExport`). In this scenario, both graph ONNX graph inputs and initializers have matching name with `.` in it (e.g. `linear.weight`)
* `rename_initializer=True` is meant to be used by `torch.onnx.dynamo_export` with a non-publicly-supported FX tracer called `FXSymbolicTracer`. This tracer lifts the FX graph initializers as inputs before FX->ONNX start, and because of this, the initializer names must be valid python identifiers (meaning `.` are not supported argument name and must be replaced by `_` or similar). This causes the graph inputs to have names with `_` (e.g. `linear_weight`) while the initializers have `.` (e.g. `linear.weight`) in their name. This flag resolves this mismatch by replacing `.` by `_` when saving the ONNX proto (`save_model_with_external_data`).
* This PR also adds unit tests for numerical validation against pytorch eager for onnx export using dynamo-based fx tracer and fake mode enabled. (There are already tests for export with fx symbolic tracer with fake mode)

Commit:
Fix initializer naming at torch.onnx.ExportOutput.save_model_with_external_data

This PR also adds unit tests for numerical validation with fake mod onnx export

[ghstack-poisoned]

Commit:
Update on "Fix initializer naming at torch.onnx.ExportOutput.save_model_with_external_data"


This PR also adds unit tests for numerical validation with fake mod onnx export

[ghstack-poisoned]

Commit:
Update on "Fix initializer naming at torch.onnx.ExportOutput.save_model_with_external_data"


This PR also adds unit tests for numerical validation with fake mod onnx export

[ghstack-poisoned]

BowenBao commented on torch/onnx/_internal/fx/serialization.py:
```diff
@@ -126,30 +128,27 @@ def save_model_with_external_data(
             #       "transformer_attention_weight" could be "attention_weight".
             #  2.  Set "tensor" as the initializer of the matched ONNX input.
             #      E.g., "tensor" is stored as the initializer of "attention_weight".
             # Step 1 is required because sometimes, tensor names are stored with prefix the dictionary
```
omg another hack

I believe this is not needed for dynamo path fake export.

BowenBao replied:
In fact I think this logic can be put under `rename_initializer` flag too.

thiagocrepaldi replied:
good catch

BowenBao commented on torch/onnx/_internal/fx/serialization.py:
```diff
@@ -111,6 +112,7 @@ def save_model_with_external_data(
         onnx_model: ONNX model to be saved with external initializers.
             If an input name matches a tensor loaded from "torch_load_paths",
             the tensor will be saved as that input's external initializer.
+        rename_initializer: Replaces "." by "_" for all ONNX initializer names.
```
Let's add some explanation. We want to call out this is only needed by fx_symbolic_tracer path.

thiagocrepaldi replied:
There was one at https://github.com/pytorch/pytorch/pull/105002/files#diff-0c7a4333620a22a5c3e5315e30272b59fb7a11b393cb42f8255070bedeb02738L125, but I will add in the docstring too (in fact, I had added before arghh)

BowenBao commented on test/onnx/test_fx_to_onnx_with_onnxruntime.py:
```diff
@@ -821,6 +826,116 @@ def create_pytorch_only_extra_kwargs():
+
+        Arguments:
+            model_name: Name of the model. It used to name temporary files.
+            create_model: A function that creates a model. It should always create the same model.
```
Does `same model` imply parameter values are also the same?

thiagocrepaldi replied:
Yes, because `create_model` and `create_args` are called twice in this helper
The first time it is called outside FakeTensorMode and real model and input are saved for the numerical validation.
The second time it is called within `enable_fake_mode` for the actual FX tracing.

In the end, `ExportOutput.save` "merges" the ONNX model without initializers with the initializers found in the first call to `create_model`/`create_args`.

BowenBao replied:
But the implementation randomly initializes parameter values, so parameter values are not the same for the models from two invokation.

Not that it matters for the test though.

thiagocrepaldi replied:
Agreed, it happens and it doesnt matter for us :)
I will remove the "same model" from docstring to kill the ambiguity.

BowenBao commented on test/onnx/test_fx_to_onnx_with_onnxruntime.py:
```diff
@@ -821,6 +826,116 @@ def create_pytorch_only_extra_kwargs():
+            model_name: Name of the model. It used to name temporary files.
+            create_model: A function that creates a model. It should always create the same model.
+            create_args: A function that creates random input arguments for the model.
+            create_pytorch_only_kwargs: A function that creates kwargs for calling PyTorch model with real tensors.
```
`create_pytorch_only_kwargs` -> `create_kwargs`

No longer a limitation for us.

BowenBao commented on test/onnx/test_fx_to_onnx_with_onnxruntime.py:
```diff
@@ -821,6 +826,116 @@ def create_pytorch_only_extra_kwargs():
+        create_args: Callable,
+        create_pytorch_only_kwargs: Callable,
+    ):
+        """Test helper for large-scale exporter.
```
nit: name diff w/ method

BowenBao commented on test/onnx/test_fx_to_onnx_with_onnxruntime.py:
```diff
@@ -821,6 +826,116 @@ def create_pytorch_only_extra_kwargs():
+                args = create_args()
+                kwargs = create_pytorch_only_kwargs()
+                # Original outputs.
+
```
```suggestion
```
nitnit empty line since the comment applies to the line after.

BowenBao commented on test/onnx/test_fx_to_onnx_with_onnxruntime.py:
```diff
@@ -821,6 +826,116 @@ def create_pytorch_only_extra_kwargs():
+                # ORT outputs.
+                args_not_none = export_output.adapt_torch_inputs_to_onnx(*args)
+                # Drop Parameters and buffers added by fx_serialization.save_model_with_external_data
+                args_not_none = args_not_none[: len(args) - len(kwargs)]
```
I think this line not necessary?

thiagocrepaldi replied:
nope, not needed. I think i messed up my branch at some point today lol
a few lines look like older commits

BowenBao commented on test/onnx/test_fx_to_onnx_with_onnxruntime.py:
Great first step. Can we parameterize `fake_export` just like other configurations like `dynamic_shape`, `op_level_debug`, etc., such that we have better test coverage, and test coverage is auto extended when new tests are added.

thiagocrepaldi replied:
Both are not fully supported yet. They are TBD in next tasks

Commit:
Update on "Fix initializer naming at torch.onnx.ExportOutput.save_model_with_external_data"


This PR is only relevant for the Fake tensor Mode ONNX export. For the conventional export, everything is unchanged.

* An optional `rename_initializer=False` argument is added to an internal function `torch/onnx/_internal/fx/serialization.py::save_model_with_external_data` which is used by the public API `ExportOutput.save`. 
* The default behavior (`rename_initializer=False`) is meant to be used by public API `torch.onnx.dynamo_export` with the default Dynamo-based FX tracer (`DynamoExport`). In this scenario, both graph ONNX graph inputs and initializers have matching name with `.` in it (e.g. `linear.weight`)
* `rename_initializer=True` is meant to be used by `torch.onnx.dynamo_export` with a non-publicly-supported FX tracer called `FXSymbolicTracer`. This tracer lifts the FX graph initializers as inputs before FX->ONNX start, and because of this, the initializer names must be valid python identifiers (meaning `.` are not supported argument name and must be replaced by `_` or similar). This causes the graph inputs to have names with `_` (e.g. `linear_weight`) while the initializers have `.` (e.g. `linear.weight`) in their name. This flag resolves this mismatch by replacing `.` by `_` when saving the ONNX proto (`save_model_with_external_data`).
* This PR also adds unit tests for numerical validation against pytorch eager for onnx export using dynamo-based fx tracer and fake mode enabled. (There are already tests for export with fx symbolic tracer with fake mode)

[ghstack-poisoned]

Commit:
Update on "Fix initializer naming at torch.onnx.ExportOutput.save_model_with_external_data"


This PR is only relevant for the Fake tensor Mode ONNX export. For the conventional export, everything is unchanged.

* An optional `rename_initializer=False` argument is added to an internal function `torch/onnx/_internal/fx/serialization.py::save_model_with_external_data` which is used by the public API `ExportOutput.save`. 
* The default behavior (`rename_initializer=False`) is meant to be used by public API `torch.onnx.dynamo_export` with the default Dynamo-based FX tracer (`DynamoExport`). In this scenario, both graph ONNX graph inputs and initializers have matching name with `.` in it (e.g. `linear.weight`)
* `rename_initializer=True` is meant to be used by `torch.onnx.dynamo_export` with a non-publicly-supported FX tracer called `FXSymbolicTracer`. This tracer lifts the FX graph initializers as inputs before FX->ONNX start, and because of this, the initializer names must be valid python identifiers (meaning `.` are not supported argument name and must be replaced by `_` or similar). This causes the graph inputs to have names with `_` (e.g. `linear_weight`) while the initializers have `.` (e.g. `linear.weight`) in their name. This flag resolves this mismatch by replacing `.` by `_` when saving the ONNX proto (`save_model_with_external_data`).
* This PR also adds unit tests for numerical validation against pytorch eager for onnx export using dynamo-based fx tracer and fake mode enabled. (There are already tests for export with fx symbolic tracer with fake mode)

[ghstack-poisoned]

BowenBao commented:
@pytorchbot merge
